tests/unit/s2n_tls13_record_aead_test.c

```diff
@@ -0,0 +1,230 @@
+    "00170018001901000101010201030104001c000240"
+    "01000000000b0001b9000001b50001b0308201ac30"
+    "820115a003020102020102300d06092a864886f70d"
+    "01010b0500300e310c300a06035504031303727361"
+    "301e170d3136303733303031323335395a170d3236"
+    "303733303031323335395a300e310c300a06035504"
+    "03130372736130819f300d06092a864886f70d0101"
+    "01050003818d0030818902818100b4bb498f827930"
+    "3d980836399b36c6988c0c68de55e1bdb826d3901a"
+    "2461eafd2de49a91d015abbc9a95137ace6c1af19e"
+    "aa6af98c7ced43120998e187a80ee0ccb0524b1b01"
+    "8c3e0b63264d449a6d38e22a5fda43084674803053"
+    "0ef0461c8ca9d9efbfae8ea6d1d03e2bd193eff0ab"
+    "9a8002c47428a6d35a8d88d79f7f1e3f0203010001"
+    "a31a301830090603551d1304023000300b0603551d"
+    "0f0404030205a0300d06092a864886f70d01010b05"
+    "000381810085aad2a0e5b9276b908c65f73a726717"
+    "0618a54c5f8a7b337d2df7a594365417f2eae8f8a5"
+    "8c8f8172f9319cf36b7fd6c55b80f21a0301515672"
+    "6096fd335e5e67f2dbf102702e608ccae6bec1fc63"
+    "a42a99be5c3eb7107c3c54e9b9eb2bd5203b1c3b84"
+    "e0a8b2f759409ba3eac9d91d402dcc0cc8f8961229"
+    "ac9187b42b4de100000f000084080400805a747c5d"
+    "88fa9bd2e55ab085a61015b7211f824cd484145ab3"
+    "ff52f1fda8477b0b7abc90db78e2d33a5c141a0786"
+    "53fa6bef780c5ea248eeaaa785c4f394cab6d30bbe"
+    "8d4859ee511f602957b15411ac027671459e46445c"
+    "9ea58c181e818e95b8c3fb0bf3278409d3be152a3d"
+    "a5043e063dda65cdf5aea20d53dfacd42f74f31400"
+    "00209b9b141d906337fbd2cbdce71df4deda4ab42c"
+    "309572cb7fffee5454b78f071816"; /* includes last byte for content type */
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Test s2n_tls13_aead_aad_init() */
+    {
+        s2n_stack_blob(aad, S2N_TLS13_AAD_LEN, S2N_TLS13_AAD_LEN);
+        struct s2n_stuffer associated_data_stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_init(&associated_data_stuffer, &aad));
+        EXPECT_SUCCESS(s2n_tls13_aead_aad_init(662, 12, &associated_data_stuffer));
+        S2N_BLOB_FROM_HEX(expected_aad, "17030302a2");
+        S2N_BLOB_EXPECT_EQUAL(expected_aad, aad);
+        S2N_BLOB_FREE(expected_aad);
+
+        /* record length 16640 should be valid */
+        EXPECT_SUCCESS(s2n_stuffer_rewrite(&associated_data_stuffer));
+        EXPECT_SUCCESS(s2n_tls13_aead_aad_init(16628, 12, &associated_data_stuffer));
+
+        /* record length 16641 should be invalid */
+        EXPECT_SUCCESS(s2n_stuffer_rewrite(&associated_data_stuffer));
+        EXPECT_FAILURE_WITH_ERRNO(s2n_tls13_aead_aad_init(16629, 12, &associated_data_stuffer), S2N_ERR_RECORD_LIMIT);
+
+        /* Test other failure cases */
+        EXPECT_SUCCESS(s2n_stuffer_rewrite(&associated_data_stuffer));
+        EXPECT_FAILURE(s2n_tls13_aead_aad_init(16629, 12, NULL));
+
+        EXPECT_SUCCESS(s2n_stuffer_rewrite(&associated_data_stuffer));
+        EXPECT_FAILURE(s2n_tls13_aead_aad_init(16628, 0, &associated_data_stuffer));
+
+        EXPECT_FAILURE(s2n_tls13_aead_aad_init(-1, 0, &associated_data_stuffer));
```

Comment:
`/* No AAD should be invalid */`

`/* 0-length tag should be invalid */`

`/* invalid record length (-1) should be invalid */`

etc.

Reply:
thanks, will add them!